barretenberg/cpp/src/barretenberg/vm/avm/trace/bytecode_trace.cpp

```diff
@@ -3,20 +3,60 @@
-AvmBytecodeTraceBuilder::AvmBytecodeTraceBuilder(const std::vector<AvmContractBytecode>& all_contracts_bytecode)
-    : all_contracts_bytecode(all_contracts_bytecode)
-{}
+using Point = grumpkin::g1::affine_element;
+
+/**************************************************************************************************
+ *                                   HELPERS
+ **************************************************************************************************/
+// TODO: Handling of point at infinity is still a bb <> ts pain point, we need to make sure we are consistent
+FF AvmBytecodeTraceBuilder::compute_address_from_instance(ContractInstanceHint contract_instance)
+{
+    FF salted_initialization_hash = poseidon2::hash({ PARTIAL_ADDRESS,
+                                                      contract_instance.salt,
+                                                      contract_instance.initialisation_hash,
+                                                      contract_instance.deployer_addr });
+    FF partial_address =
+        poseidon2::hash({ PARTIAL_ADDRESS, contract_instance.contract_class_id, salted_initialization_hash });
+    // Double check noir needs to hash infinity flag...
+    std::vector<FF> public_keys_hash_fields = contract_instance.public_keys.to_fields();
+    std::vector<FF> public_key_hash_vec{ PUBLIC_KEYS_HASH };
+    for (size_t i = 0; i < public_keys_hash_fields.size(); i += 2) {
+        public_key_hash_vec.push_back(public_keys_hash_fields[i]);
+        public_key_hash_vec.push_back(public_keys_hash_fields[i + 1]);
+        // Is it guaranteed we wont get a point at infinity here?
+        public_key_hash_vec.push_back(FF::zero());
+    }
+    FF public_keys_hash = poseidon2::hash({ public_key_hash_vec });
+    FF h = poseidon2::hash({ CONTRACT_ADDRESS_V1, public_keys_hash, partial_address });
+    // This is safe since BN254_Fr < GRUMPKIN_Fr so we know there is no modulo reduction
+    grumpkin::fr h_fq = grumpkin::fr(h);
+    return (grumpkin::g1::affine_one * h_fq + contract_instance.public_keys.incoming_viewing_key).x;
+}
+
+FF AvmBytecodeTraceBuilder::compute_contract_class_id(FF artifact_hash,
+                                                      FF private_fn_root,
+                                                      FF public_bytecode_commitment)
+{
+    return poseidon2::hash({ CONTRACT_LEAF, artifact_hash, private_fn_root, public_bytecode_commitment });
+}
 
-std::vector<FF> AvmBytecodeTraceBuilder::pack_bytecode(const std::vector<uint8_t>& contract_bytes)
+std::vector<FF> AvmBytecodeTraceBuilder::encode_bytecode(const std::vector<uint8_t>& contract_bytes)
 {
     // To make from_buffer<uint256_t> work properly, we need to make sure the contract is a multiple of 31 bytes
     // Otherwise we will end up over-reading the buffer
-    size_t padding = 31 - (contract_bytes.size() % 31);
+    size_t padded_size = 31 * ((contract_bytes.size() + 30) / 31);
```

Comment:
What is this change doing? Maybe comment this line explaining exactly what it's doing

Reply:
Ah it's not doing a mod anymore. Just calculating the number of fields and then multiplying by 31 to determine the #bytes needed.

We're always adding a +30 here? And then relying on integer division to toss any remainder so it has the same effect as the pre-diff code?

Honestly it was more readable before this change :sweat_smile:

Reply:
Haha yeah but @jeanmon convinced me this is a nicer way to handle padding!